File: packages/core/src/message/MessageSerializer.ts
import type { BridgeRequest } from '../types.js';
import { generateId } from '../util/uuid.js';
import { ErrorCode, RynBridgeError } from '../errors.js';

export class MessageSerializer {
  private readonly version: string;

  constructor(version: string) {
    this.version = version;
  }

  createRequest(
    module: string,
    action: string,
    payload: Record<string, unknown> = {},
  ): BridgeRequest {
    return {
      id: generateId(),
      module,
      action,
      payload,
      version: this.version,
    };
  }

  serialize(request: BridgeRequest): string {
    try {
      return JSON.stringify(request);
    } catch (error) {
      throw new RynBridgeError(
        ErrorCode.SERIALIZATION_ERROR,
        'Failed to serialize message',
        { originalError: String(error) },
      );
    }
  }
}

File: android/core/src/main/kotlin/com/devryner/rynbridge/core/MessageSerializer.kt
package com.devryner.rynbridge.core

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

class MessageSerializer(private val version: String = "0.1.0") {

    private val json = Json { encodeDefaults = true }

    fun createRequest(
        module: String,
        action: String,
        payload: Map<String, BridgeValue> = emptyMap()
    ): BridgeRequest {
        return BridgeRequest(
            id = UUID.randomUUID().toString().lowercase(),
            module = module,
            action = action,
            payload = payload,
            version = version
        )
    }

    fun createResponse(
        id: String,
        status: ResponseStatus,
        payload: Map<String, BridgeValue> = emptyMap(),
        error: BridgeErrorData? = null
    ): BridgeResponse {
        return BridgeResponse(id = id, status = status, payload = payload, error = error)
    }

    fun serialize(request: BridgeRequest): String {
        try {
            return json.encodeToString(request)
        } catch (e: RynBridgeError) {
            throw e
        } catch (e: Exception) {
            throw RynBridgeError(
                code = ErrorCode.SERIALIZATION_ERROR,
                message = "Failed to serialize request: ${e.message}"
            )
        }
    }

    fun serialize(response: BridgeResponse): String {
        try {
            return json.encodeToString(response)
        } catch (e: RynBridgeError) {
            throw e
        } catch (e: Exception) {
            throw RynBridgeError(
                code = ErrorCode.SERIALIZATION_ERROR,
                message = "Failed to serialize response: ${e.message}"
            )
        }
    }
}
